HelpSource/Reference/ServerPluginAPI.schelp

```diff
@@ -0,0 +1,189 @@
+
+## code::IN0(index)::
+|| A single sample of control-rate input as a float, at the given index.
+
+## code::OUT(index)::
+|| A single block of audio-rate output as a float* at the given index.
+
+## code::OUT0(index)::
+|| A single sample of control-rate input as a float, at the given index.
+
+## code::INRATE(index)::
+|| Get the rate of a given input index. This will be one of the four rates.
+::
+
+section:: Unary operators
+
+definitionlist::
+## code::bool sc_isnan(float/double x)::
+|| Is code::x:: NaN?
+
+## code::bool sc_isfinite(float/double x)::
+|| Is code::x:: finite?
+
+## code::int32 sc_grayCode(int32 x)::
+|| Convert binary to Gray code.
+::
+
+The following unary functions are available for both float32 and float64, and are the same as in sclang (minus the "sc_" prefixes):
+
+code::
+sc_log2
+sc_log10
+sc_midicps
+sc_cpsmidi
+sc_midiratio
+sc_ratiomidi
+sc_octcps
+sc_cpsoct
+sc_ampdb
+sc_dbamp
+sc_cubed
+sc_sqrt
+sc_hanwindow
+sc_welwindow
+sc_triwindow
+sc_rectwindow
+sc_scurve
+sc_ramp
+sc_sign
+sc_distort
+sc_softclip
+sc_trunc
+sc_ceil
+sc_floor
+sc_reciprocal
+sc_frac
+::
+
+The following unary functions are available for both float32 and float64, but have no sclang equivalent:
+
+definitionlist::
+## code::zapgremlins(x)::
+|| Replaces NaNs, infinities, very large and very small numbers (including denormals) with zero. This is useful in ugen feedback to safeguard from pathological behavior. (Note lack of sc_ prefix.)
+
+## code::sc_bitriwindow::
+|| Alternative to code::sc_triwindow:: using absolute value.
+
+## code::sc_scurve0(x)::
+|| Same as code::sc_scurve::, but assumes that code::x:: is in the interval [0, 1].
+
+## code::sc_distortneg(x)::
+|| A one-sided distortion function. Same as code::distort:: for code::x > 0::, and the identity function for code::x <= 0::.
+
+## code::taylorsin(x)::
+|| Taylor series approximation of code::sin(x):: out to code::x**9 / 9!::. (Note lack of sc_ prefix.)
+
+## code::sc_lg3interp(x1, a, b, c, d)::
+|| Cubic Lagrange interpolator.
+
+## code::sc_CalcFeedback(delaytime, decaytime)::
+|| Determines the feedback coefficient for a feedback comb filter with the given delay and decay times.
+
+## code::sc_wrap1(x)::
+|| Same as code::sc_wrap::, but wraps only once.
+
+## code::sc_fold1(x)::
+|| Same as code::sc_fold::, but folds only once.
+::
+
+section:: Unroll macros
+
+Audio-rate ugens often use loops that look like this:
+
+code::
+float* in = IN(0);
+float* out = OUT(0);
+for (int i = 0; i < inNumSamples; i++) {
+	out[i] = in[i] * 0.5;
+}
+::
+
+The indexing operations code::out[i]:: and code::in[i]:: invoke pointer addition at every sample. We can speed this up by using increments rather than addition, and speed it up further by using pre-increments, not post-increments (so there is no need for the CPU to preserve the old value). The result is the following:
```

Comment:
this part is pretty wrong, 20 years ago on ppc, this might have been the case, but not these days anymore